File: frontend/services/spellcheck.ts
import { api } from "./api";
import type {
  SpellCheckWordResponse,
  SpellCheckTextResponse,
} from "../types/spellcheck";

export async function checkWord(
  word: string,
  maxDistance = 2,
  topN = 5
): Promise<SpellCheckWordResponse> {
  const params = new URLSearchParams({
    word,
    max_distance: String(maxDistance),
    top_n: String(topN),
  });
  return api.get<SpellCheckWordResponse>(`/spellcheck/word?${params}`);
}

export async function checkText(
  text: string,
  maxDistance = 2,
  topN = 5
): Promise<SpellCheckTextResponse> {
  const params = new URLSearchParams({
    max_distance: String(maxDistance),
    top_n: String(topN),
  });
  const API_URL = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000";
  const response = await fetch(`${API_URL}/spellcheck/check-text?${params}`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ text }),
  });
  if (!response.ok) {
    throw new Error(`Failed to check text: ${response.status}`);
  }
  return response.json();
}

export async function addWordToDictionary(word: string): Promise<{ status: string; word: string }> {
  return api.post<{ status: string; word: string }>("/dictionary/add", { word });
}

File: backend/app/api/routes/spellcheck.py
from fastapi import APIRouter, Body, HTTPException, Query, Request

from app.api.schemas.spellcheck import (
    LevenshteinTableResponse,
    SpellCheckCandidateResponse,
    SpellCheckIssueResponse,
    SpellCheckTextRequest,
    SpellCheckTextResponse,
    SpellCheckWordResponse,
)
from app.core.levenshtein import compute_distance_table
from app.services.spellcheck_service import SpellCheckService

router = APIRouter(tags=["spellcheck"])

def _get_spellcheck_service(request: Request) -> SpellCheckService:
    dictionary_service = getattr(request.app.state, "dictionary_service", None)

    if dictionary_service is None or not dictionary_service.is_loaded:
        raise HTTPException(503, "Dictionary service is not loaded")

    return SpellCheckService(dictionary_service)

@router.get("/spellcheck/word", response_model=SpellCheckWordResponse)
async def check_word(
    request: Request,
    word: str = Query(..., min_length=1, max_length=64),
    max_distance: int = Query(2, ge=0, le=5),
    top_n: int = Query(5, ge=1, le=20),
) -> SpellCheckWordResponse:
    service = _get_spellcheck_service(request)

    is_valid = service.is_valid_word(word)

    suggestions: list[SpellCheckCandidateResponse] = []

    if not is_valid:
        candidates = service.find_words_within_distance(
            word, max_distance=max_distance, limit=top_n
        )

        suggestions = [
            SpellCheckCandidateResponse(
                word=candidate.word,
                distance=candidate.distance,
                frequency=candidate.frequency,
            )
            for candidate in candidates
        ]

    return SpellCheckWordResponse(
        word=word,
        is_valid=is_valid,
        suggestions=suggestions,
    )

@router.post("/spellcheck/check-text", response_model=SpellCheckTextResponse)
async def check_text(
    request: Request,
    payload: SpellCheckTextRequest = Body(...),
    max_distance: int = Query(2, ge=0, le=5),
    top_n: int = Query(5, ge=1, le=20),
) -> SpellCheckTextResponse:
    service = _get_spellcheck_service(request)

    issues = service.check_text(
        payload.text, max_distance=max_distance, top_n=top_n
    )

    return SpellCheckTextResponse(
        issues=[
            SpellCheckIssueResponse(
                word=issue.word,
                start=issue.start,
                end=issue.end,
                suggestions=[
                    SpellCheckCandidateResponse(
                        word=candidate.word,
                        distance=candidate.distance,
                        frequency=candidate.frequency,
                    )
                    for candidate in issue.suggestions
                ],
            )
            for issue in issues
        ],
        issue_count=len(issues),
    )

@router.get("/levenshtein/table", response_model=LevenshteinTableResponse)
async def levenshtein_table(
    source: str = Query(..., max_length=64),
    target: str = Query(..., max_length=64),
) -> LevenshteinTableResponse:
    result = compute_distance_table(source, target)

    return LevenshteinTableResponse(
        source=result.source,
        target=result.target,
        table=result.table,
        distance=result.distance,
    )